sdk/wit/miden.wit

```diff
@@ -0,0 +1,132 @@
+
+    /// A commitment to a non-fungible asset.
+    ///
+    /// A non-fungible asset consists of 4 field elements which are computed by hashing asset data
+    /// (which can be of arbitrary length) to produce: [d0, d1, d2, d3].  We then replace d1 with the
+    /// faucet_id that issued the asset: [d0, faucet_id, d2, d3]. We then set the most significant bit
+    /// of the most significant element to ZERO.
+    type non-fungible-asset = word;
+
+    /// A fungible or a non-fungible asset.
+    ///
+    /// All assets are encoded using a single word (4 elements) such that it is easy to determine the
+    /// type of an asset both inside and outside Miden VM. Specifically:
+    ///   Element 1 will be:
+    ///    - ZERO for a fungible asset
+    ///    - non-ZERO for a non-fungible asset
+    ///   The most significant bit will be:
+    ///    - ONE for a fungible asset
+    ///    - ZERO for a non-fungible asset
+    ///
+    /// The above properties guarantee that there can never be a collision between a fungible and a
+    /// non-fungible asset.
+    ///
+    /// The methodology for constructing fungible and non-fungible assets is described below.
+    ///
+    /// # Fungible assets
+    /// The most significant element of a fungible asset is set to the ID of the faucet which issued
+    /// the asset. This guarantees the properties described above (the first bit is ONE).
+    ///
+    /// The least significant element is set to the amount of the asset. This amount cannot be greater
+    /// than 2^63 - 1 and thus requires 63-bits to store.
+    ///
+    /// Elements 1 and 2 are set to ZERO.
+    ///
+    /// It is impossible to find a collision between two fungible assets issued by different faucets as
+    /// the faucet_id is included in the description of the asset and this is guaranteed to be different
+    /// for each faucet as per the faucet creation logic.
+    ///
+    /// # Non-fungible assets
+    /// The 4 elements of non-fungible assets are computed as follows:
+    /// - First the asset data is hashed. This compresses an asset of an arbitrary length to 4 field
+    ///   elements: [d0, d1, d2, d3].
+    /// - d1 is then replaced with the faucet_id which issues the asset: [d0, faucet_id, d2, d3].
+    /// - Lastly, the most significant bit of d3 is set to ZERO.
+    ///
+    /// It is impossible to find a collision between two non-fungible assets issued by different faucets
+    /// as the faucet_id is included in the description of the non-fungible asset and this is guaranteed
+    /// to be different as per the faucet creation logic. Collision resistance for non-fungible assets
+    /// issued by the same faucet is ~2^95.
+    variant asset {
+        fungible(fungible-asset),
+        non-fungible(non-fungible-asset),
+    }
+
+    /// Inputs of the currently executed note (4 words = 16 felts)
+    type note-inputs = tuple<felt, felt, felt, felt, felt, felt, felt, felt, felt, felt, felt, felt, felt, felt, felt, felt>;
```

Comment:
The inputs are variably-sized, i.e. there are only as many items in the inputs as there are actual inputs. I would make this a list/array of `felt`, with a note that there are never more than 16 inputs provided.

Reply:
Agree. I modeled it way too closely to the existing kernel MASM API. Fixed in https://github.com/0xPolygonMiden/compiler/pull/79/commits/c0d3aec94ce4665bb7594a4279d875e56c1a509d